text/0431-sagemaker-l2-endpoint.md

```diff
@@ -0,0 +1,1150 @@
+five containers that process requests for inferences on data. You use an inference pipeline to
+define and deploy any combination of pretrained Amazon SageMaker built-in algorithms and your own
+custom algorithms packaged in Docker containers. You can use an inference pipeline to combine
+preprocessing, predictions, and post-processing data science tasks. Inference pipelines are fully
+managed. To define an inference pipeline, you can provide additional containers for your model via
+the `extraContainers` property:
+
+```typescript fixture=with-assets
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const model = new sagemaker.Model(this, 'InferencePipelineModel', {
+  container: {
+    image: image1, modelData: modelData1
+  },
+  extraContainers: [
+    { image: image2, modelData: modelData2 },
+    { image: image3, modelData: modelData3 }
+  ],
+});
+```
+
+## Model Hosting
+
+Amazon SageMaker provides model hosting services for model deployment. Amazon SageMaker provides an
+HTTPS endpoint where your machine learning model is available to provide inferences.
+
+### Endpoint Configuration
+
+In this configuration, you identify one or more models to deploy and the resources that you want
+Amazon SageMaker to provision. You define one or more production variants, each of which identifies
+a model. Each production variant also describes the resources that you want Amazon SageMaker to
+provision. This includes the number and type of ML compute instances to deploy. If you are hosting
+multiple models, you also assign a variant weight to specify how much traffic you want to allocate
+to each model. For example, suppose that you want to host two models, A and B, and you assign
+traffic weight 2 for model A and 1 for model B. Amazon SageMaker distributes two-thirds of the
+traffic to Model A, and one-third to model B:
+
+```typescript fixture=with-assets
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const endpointConfig = new sagemaker.EndpointConfig(this, 'EndpointConfig', {
+  productionVariant: {
+    model: modelA,
+    variantName: 'modelA',
+    initialVariantWeight: 2.0,
+  },
+  extraProductionVariants: [{
+    model: modelB,
+    variantName: 'variantB',
+    initialVariantWeight: 1.0,
+  }]
+});
+```
+
+### Endpoint
+
+If you create an endpoint from an `EndpointConfig`, Amazon SageMaker launches the ML compute
+instances and deploys the model or models as specified in the configuration. To get inferences from
+the model, client applications send requests to the Amazon SageMaker Runtime HTTPS endpoint. For
+more information about the API, see the
+[InvokeEndpoint](https://docs.aws.amazon.com/sagemaker/latest/dg/API_runtime_InvokeEndpoint.html)
+API. Defining an endpoint requires at minimum the associated endpoint configuration:
+
+```typescript fixture=with-endpoint-config
+import * as sagemaker from '@aws-cdk/aws-sagemaker';
+
+const endpoint = new sagemaker.Endpoint(this, 'Endpoint', { endpointConfig });
```

Comment:
Is it valuable to have `EndpointConfig` as its own separate visible resource in the first place? Why don't I just directly declare an `Endpoint` which has 1..* models?

Reply:
Reopening to say that the related comment convinces me that having an `EndpointConfig` L2 makes sense. However, we should upstream that reasoning here so that our users will understand the use case as well.

Reply:
In rev2, I've mentioned that an `EndpointConfig` can be used to provision one or more `Endpoint` resources.